src/merkle_tree.rs

```diff
@@ -0,0 +1,516 @@
+
+impl MerkleTree {
+    /// Create a new MerkleTree from the provided items.
+    /// Each item should be representable as bytes.
+    /// The creation will fail if the items list is empty.
+    ///
+    /// # Examples
+    /// ```
+    /// use merkle_tree::MerkleTree;
+    ///
+    /// let items = vec!["In a hole in the ground", "there lived a hobbit."];
+    /// let merkle_tree = MerkleTree::build(&items).unwrap();
+    /// ```
+    pub fn build<T: AsRef<[u8]>>(items: &[T]) -> Option<Self> {
+        if items.is_empty() {
+            return None;
+        }
+
+        let leaves: Vec<Hash> = items.iter().map(|item| Self::hash(item.as_ref())).collect();
+
+        let levels = Self::construct_levels(leaves);
+
+        Some(Self { levels })
+    }
+
+    /// Insert a new item into the Merkle tree.
+    /// The tree will be updated to include the new item's hash.
+    ///
+    /// # Examples
+    /// ```
+    /// use merkle_tree::MerkleTree;
+    /// let items = vec!["In a hole in the ground", "there lived a hobbit."];
+    /// let mut merkle_tree = MerkleTree::build(&items).unwrap();
+    ///
+    /// merkle_tree.insert(&"Gandalf the Grey");
+    /// ```
+    pub fn insert<T: AsRef<[u8]>>(&mut self, item: &T) {
+        let mut leaves = self.levels[0].clone();
+        leaves.push(Self::hash(item.as_ref()));
+
+        let levels = Self::construct_levels(leaves);
+
+        self.levels = levels;
+    }
+
+    fn construct_levels(leaves: Vec<Hash>) -> Vec<Vec<Hash>> {
+        let total_height = Self::tree_height(leaves.len());
+
+        let mut levels = Vec::with_capacity(total_height + 1);
+
+        levels.push(leaves);
+
+        while let Some(level) = Self::merkle_parent_level(levels.last().unwrap()) {
+            levels.push(level);
+        }
+
+        levels
+    }
+
+    fn tree_height(items: usize) -> usize {
+        (items as f64).log2().ceil() as usize
+    }
+
+    /// Computes the parent hash for the concatenation of the children hashes.
+    fn merkle_parent(children: &[Hash]) -> Hash {
+        let mut children_vector = children.to_vec();
+        children_vector.sort();
+        Self::hash(children_vector.as_flattened())
+    }
+
+    /// Creates the parent level for the given level.
+    /// If the level has an odd number of hashes, the last hash is duplicated.
+    fn merkle_parent_level(level: &Vec<Hash>) -> Option<Vec<Hash>> {
+        // Is root, return None.
+        if level.len() == 1 {
+            return None;
+        }
+
+        // If the number of leafs is odd, duplicate the last leaf.
+        let mut parent_level = level.clone();
+
+        if level.len() % 2 == 1 {
+            parent_level.extend(parent_level.last().cloned())
+        }
+
+        Some(
+            parent_level
+                .chunks_exact(2)
+                .map(Self::merkle_parent)
+                .collect(),
+        )
+    }
+
+    /// Computes the Merkle root hash for the provided leaf hashes.
+    pub fn root(&self) -> Hash {
```

Comment:
could return an `Option<Hash>` to skip the unwraps

Reply:
Good point. Will do.